### DISCUSSION TITLE
How to access Plotlys "extendTraces" API?

### DISCUSSION THREAD

**Discussion by tscheburaschka** [zauberzeug/nicegui#2660] [Q&A]
### Question

Hi,
I'm trying to use nicegui for a live figure with frequent updates triggered by the server.
Until now I used Plotly Dash for this and here they expose the [`Plotly.extendTraces`](https://plotly.com/javascript/plotlyjs-function-reference/#plotlyextendtraces) API as [extendData](https://dash.plotly.com/dash-core-components/graph) in the Python classes.
The advantage of this method is a dramatically reduced traffic for updates that add only a few points to a large graph.
I tried to use this API with a naive call to nicegui's generic `runMethod` like 
`plot.run_method("Plotly.extendTraces", {"y": [[3, 2], [5, 1]]}, [0, 1], 10)`
but this fails, because the id of the plotly div is sent as `4` (int) where it should be `'c4'` (found through inspection on the client page). But even when fixing the id, things get messy, because the server state of the figure is now out of sync with the client and any subsequent call to `update` transmits the whole figure. The server could avoid calling `update` ever again but obviously, things are more complicated than this.

Sooo, is there any hope? More precisely, what path would you recommend to follow on?

Thanks for any help!

**Comment by falkoschindler** [comment]
That's a good question, @tscheburaschka!

I don't see an obvious way to avoid this problem: You either update the client state via `run_method` calls and get client and server out of sync, or you re-transmit the full state again and again.

For elements like `ui.scene` we actually implemented an alternative way: The server doesn't send the whole state at once, but a number of actions like "create sphere", "move sphere", "move camera", that can be extended later. But this is quite complex and doesn't scale well, because you kind of need to mirror the JavaScript API of the underlying library.

`ui.log` and `ui.line_plot` are even better examples: You can `push` log lines or data points and NiceGUI will make sure that client and server are equally updated - without re-sending the complete state.

Maybe we can find such a lightweight solution for `ui.plotly` as well. Could a method like `extend_traces()` manipulate the figure on the server and update the client via `run_method`? Any suggestions are very welcome!

**Comment by johanfforsberg** [comment]
I just ran into the same problem.

For my use case I went down the rabbit hole and used `run_javascript` like this:
```python
ui.run_javascript(f"Plotly.extendTraces(document.querySelector(''.js-plotly-plot''), {update_json}, {traces_json}")
```
It works, but is obviously hacky since it assumes that there's only one plot on the page. I didn't manage to find it using `getElement()`.

Having a helper method on the `ui.plotly` element to do this properly would be a good addition.

**Reply by falkoschindler** [reply]
Hi @johanfforsberg,

I guess the problem with `getElement` is that it returns the Vue component, not the HTML DOM element. You can use `getHtmlElement` instead:

```py
plot = ui.plotly({'data': [{'x': [], 'y': []}]})

def extend() -> None:
    update = {'x': [[time.time()]], 'y': [[random()]]}
    ui.run_javascript(f'Plotly.extendTraces(getHtmlElement({plot.id}), {json.dumps(update)}, [0])')

ui.timer(0.5, extend)
```

Regarding a helper method on `ui.plotly`, we're open for feature requests or PRs.
The main design question would be whether such a helper should *only* push the delta to the client (fast, but the server-side `fig.data` drifts, so a later `plot.update()` snaps back to the old state), or also mirror the change into `fig.data` to keep both sides in sync.
The latter is nicer to use but locks the helper to scatter-style traces with `x`/`y`, and raises the question of how far to go (`prepend_traces`, `add_traces`, `delete_traces`, …).
Happy to hear your thoughts!

**Reply by johanfforsberg** [reply]
Thanks for the quick reply, `getHtmlElement` works perfectly! 

(I have only been playing around with NiceGUI for a week or so but I am very impressed with it so far. My use case is scientific control systems, so appending data to a real time plot is quite a common situation.)

I'd be happy to look into contributing something, but agree that the trick is to come up with a "general enough" solution. I suppose the internal plot should be kept updated, to minimize confusion. 

Not sure about other operations than "append". In the end, this is about performance, not really functionality as such, right? At least for me, appending is the most important use case here as it can be a huge performance boost to not upload the whole data over and over again. Prepend might mostly be useful for "backfilling" data, which is usually only done rarely anyway? Adding and deleting traces could be useful though. 

There's also the matter of plotly's different plot types, not sure where appending to a trace makes sense. E.g. imshow? There are use cases where an image would need incremental updating, but there I think it's not really about appending but replacing, since all the pixels are already present. I will play around with that.

I'd be happy with doing it the "low level" way for now, it's pretty simple and seems robust enough. But the main problem may be that it's obscure, perhaps documenting this way to do it would be a good start?

**Comment by evnchn** [comment]
@falkoschindler I think we need some unified API so that user can, if they choose to, to engage JS logic so that when our Python-side data store is updated, so is the browser, to best efforts. 

This pattern has shown itself before. Read https://github.com/zauberzeug/nicegui/discussions/5952 and you will find the story similar, just that this is Plotly, that is AG Grid. 

I feel like this is the last piece of the interactivity puzzle NiceGUI should solve.

**Reply by johanfforsberg** [reply]
It would sure be nice to have an "idiomatic" and general way to do this kind of thing.

I am fine with reaching into the internals now and then in order to do uncommon stuff, even if it's a bit ugly like building javascript strings. So far NiceGUI seems to fit quite well with the "making simple things easy and hard things possible" philosophy.

**Reply by falkoschindler** [reply]
@evnchn So you're suggesting a cross-cutting API for syncing client and server? I'm open for ideas how this could look like.

Otherwise, an additional demo at https://nicegui.io/documentation/plotly would at least avoid this specific confusion.

**Reply by evnchn** [reply]
Crosscutting API will come later

Per past practice, shall document what to do with present API first.

**Comment by evnchn** [comment]
Picking up the open design question @falkoschindler raised earlier in this thread — *"Could a method like `extend_traces()` manipulate the figure on the server and update the client via `run_method`?"* — and bringing in context from the parallel AG Grid thread in [#5952](https://github.com/zauberzeug/nicegui/discussions/5952), where the same shape of problem already landed on docs + recipe rather than a wrapper method (see [PR #5967](https://github.com/zauberzeug/nicegui/pull/5967)).

**Concrete proposal — mirror what #5967 did for AG Grid, for Plotly:**

1. A `@doc.demo` showing the `suspend_updates()` + `Plotly.extendTraces` recipe via `plot.run_method(...)` directly (no `ui.run_javascript` needed). There's a small JS-side wart on `ui.plotly` where `run_method` doesn't resolve `Plotly.*` library calls against the HTML root — worth fixing as part of the same PR so the recipe works without dropping to `ui.run_javascript`.
2. A short top-level docs page — *"Pushing deltas to elements"* — that documents the pattern once and links out to the [plotly.js function reference](https://plotly.com/javascript/plotlyjs-function-reference/) and [AG Grid grid-api reference](https://www.ag-grid.com/javascript-data-grid/grid-api/) for the menu of library methods. @johanfforsberg's point above (*"perhaps documenting this way to do it would be a good start?"*) and @peyman-vasegh's same ask in #5952 are both about discoverability, not new API: those JS-library URLs are currently nowhere in NiceGUI's Python-side docs, so the only natural path to them is reading `run_method`'s docstring.

This answers the open design question with the *lighter* side: don't add `extend_traces()` as a wrapper method (it would commit NiceGUI to mirroring plotly.js's mutation vocabulary forever — same slippery-slope concern Falko raised against the AG Grid version in #5952). Just make the existing primitive (`run_method` + `suspend_updates`) discoverable and gotcha-free for the high-value cases.

<details>
<summary>What about a framework version (`delta()` context manager + diff-translators)?</summary>

I sketched one with @evnchn: `with element.delta() as d: ...; d.apply()` plus per-element diff-translators that map "you appended to `figure.data[0].x`" → `Plotly.extendTraces(...)` automatically, so users wouldn't have to learn the JS-library method names. Concretely: 2 translators (Plotly, AG Grid), ~5 mutation patterns each, ~60–100 lines per translator, fallback to full `update()` for anything not recognized.

It's a real UX improvement — but (a) NiceGUI would own the translator vocabulary, including library-version churn when AG Grid 35 or plotly.js 3.0 reshape their methods, and (b) it's addressing the *same* discoverability problem that documentation also addresses, at much lower cost and risk.

So the suggestion is: **don't ship the translator in core for now.** If someone wants the ergonomics, it can live as an external package (third-party prototype, no API surface owned by NiceGUI, no breaking-change risk when the heuristics get a diff wrong). If field uptake is real, NiceGUI 4.0 is the natural decision point to absorb it. If it isn't, no harm done.

</details>

— Drafted by Claude Code on behalf of @evnchn.

**Comment by falkoschindler** [comment]
PR #60606 (NiceGUI 3.13) adds a Plotly.extendTraces demo and a `run_plot_method` helper. So let's close this discussion.
